### ISSUE TITLE
Handle housekeeping with sharded cluster

### ISSUE THREAD

**Issue by hackerwins** [yorkie-team/yorkie#505]
**What would you like to be added**:

We recently introduced cluster mode based on the shard key. This shard key is used to assign APIs requested by clients to servers in the cluster based on project or document.

Housekeeping operations should also apply this sharded cluster structure. We can think of two options:

1. Only the master server executes housekeeping in the cluster
2. Split housekeeping tasks per project and execute them on servers mapped by project

**Why is this needed**:

- Bugs may occur in housekeeping of sharded cluster.

**Comment by krapie** [comment]
The options that you've suggested look good to me.
I have listed some of the considerations for these options below.

1. Only the master server executes housekeeping in the cluster.
  - Implementation: Using etcd with the raft leader election algorithm will ease this implementation. We can use K8s's etcd for this operation: https://github.com/travisjeffery/kube-leaderelection.
  - Pros: It looks easy to implement.
  - Cons: As workload for housekeeping increases, housekeeping overhead on the master server will increase, which may lead to server panic or failure. 
3. Split housekeeping tasks per project and execute them on servers mapped by project.
  - Implementation: We need to map the server with the project range to housekeeping in etcd and pull this information before housekeeping. Maybe we can reference Yorkie's etcd implementation in the sync package for this task: https://github.com/yorkie-team/yorkie/blob/fdd62a95e802073584e861e1e305f93b24a53d8b/design/cluster-mode.md
  - Pros: We can distribute housekeeping workload just like document workload sharding, which will ease the overhead of servers to handle housekeeping.
  - Cons: Implementation seems to be more difficult than the first option. 

These options seem like opposite options to me, but I think they both can be implemented by using etcd.
I'll keep searching for more information on these options.

**Comment by krapie** [comment]
I have implemented leader election with K8s [lease object](https://kubernetes.io/docs/concepts/architecture/leases/), but this will make yorkie k8s dependent.

Therefore, I will change this implementation with MongoDB, where yorkie already have dependency on.

To be specific about the implementation, I will use the way of nodes try to update on the same document(same `_id`) by checking `lease_expire_at` field.

Also, to ensure that only one node becomes the leader when a lease has expired, I will use [TTL indexes](https://www.mongodb.com/docs/manual/core/index-ttl/) to remove document on expiry, preventing nodes to update the same document simultaneously.

**Comment by hackerwins** [comment]
No, it's still on hold. During housekeeping, it seems that Client Deactivation competing across nodes doesn't cause significant issues, so we haven't proceeded with this.

**Comment by hackerwins** [comment]
This was fixed by #1373
